### PR TITLE
Javadocs api and core

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR contains a lot of new javadocs for api and core

Also:

* added the missing module-name in amqp module which was preventing the javadocs compilation
* updates the javadoc plugin to latest version
* exclude some modules from javadoc build

Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

### CODE REVIEW

File: api/src/main/java/io/cloudevents/rw/CloudEventReader.java [hunk 1]
```diff
@@ -30,18 +30,24 @@
...
 
     /**
-     * Like {@link CloudEventReader#read(CloudEventWriterFactory)}, but providing a mapper for {@link io.cloudevents.CloudEventData} to be invoked when the data field is available.
+     * Read self using the provided writer factory.
+     *
+     * @param <W>           the {@link CloudEventWriter} type
+     * @param <R>           the return type of the {@link CloudEventWriter}
+     * @param writerFactory a factory that generates a visitor starting from the SpecVersion of the event
+     * @param mapper        the mapper to invoke when building the {@link CloudEventData}
+     * @return the return value returned by {@link CloudEventWriter#end()} or {@link CloudEventWriter#end(CloudEventData)}
```
Comment (matejvasek) [inline api/src/main/java/io/cloudevents/rw/CloudEventReader.java:48]: "the return value returned by", maybe replace this by "the value returned by" or "the return value of"

File: core/src/main/java/io/cloudevents/core/extensions/impl/ExtensionUtils.java [hunk 2]
```diff
@@ -17,13 +17,23 @@
...
 
-    public static IllegalArgumentException generateInvalidKeyException(String extensionName, String key) {
-        return new IllegalArgumentException(extensionName + " doesn't expect the attribute key \"" + key + "\"");
+    /**
+     * @param clazz the {@link Extension} class
+     * @param key   the invalid key
+     * @return an {@link IllegalArgumentException} when trying to access a key of the extension not existing.
+     */
+    public static IllegalArgumentException generateInvalidKeyException(Class<? extends Extension> clazz, String key) {
+        return new IllegalArgumentException(clazz.getSimpleName() + " doesn't expect the attribute key \"" + key + "\"");
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/extensions/impl/ExtensionUtils.java:36]: Wouldn't be fully qualified name better?

Comment (matejvasek) [reply core/src/main/java/io/cloudevents/core/extensions/impl/ExtensionUtils.java:36]: or the `Extension` interface could contain `name()` method for human readable name?

File: core/src/main/java/io/cloudevents/core/message/MessageReader.java [hunk 3]
```diff
@@ -19,41 +19,50 @@
...
+     * @see #read(CloudEventWriterFactory, CloudEventDataMapper)
      */
-    default <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory) throws CloudEventRWException, IllegalStateException {
+    default <W extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<W, R> writerFactory) throws CloudEventRWException, IllegalStateException {
         return read(writerFactory, CloudEventDataMapper.identity());
     }
 
     /**
-     * Like {@link MessageReader#read(CloudEventWriterFactory)}, but providing a mapper for {@link io.cloudevents.CloudEventData} to be invoked when the data field is available.
+     * Read the message as binary encoded message using the provided reader factory.
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/message/MessageReader.java:46]: I am little confused here: "provided reader factory", but the provided parameter is writer factory isn't it?

File: core/src/main/java/io/cloudevents/core/message/MessageReader.java [hunk 4]
```diff
@@ -19,41 +19,50 @@
...
+     * @param mapper        the mapper to use to map the data, if any.
+     * @throws CloudEventRWException if something went wrong during the visit.
+     * @throws IllegalStateException if the message is not in binary encoding.
      */
-    <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory, CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException, IllegalStateException;
+    <W extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<W, R> writerFactory, CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException, IllegalStateException;
 
     /**
-     * Visit the message as structured encoded event using the provided visitor
+     * Read the message as structured encoded message using the provided reader
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/message/MessageReader.java:58]: Same as above.

File: core/src/main/java/io/cloudevents/core/extensions/impl/ExtensionUtils.java (hunk 2 above)
Comment (slinkydeveloper) [reply core/src/main/java/io/cloudevents/core/extensions/impl/ExtensionUtils.java:36]: I prefer to use the FQCN in this case

File: core/src/main/java/io/cloudevents/core/message/MessageReader.java (hunk 3 above)
Comment (slinkydeveloper) [reply core/src/main/java/io/cloudevents/core/message/MessageReader.java:46]: fixed

File: core/src/main/java/io/cloudevents/core/message/MessageReader.java [hunk 5]
```diff
@@ -64,36 +73,39 @@
...
-     * @return The return value of the MessageVisitor
+     * @param <BW>           the {@link CloudEventWriter} type
+     * @param <R>           the return type of both {@link CloudEventWriter} and {@link StructuredMessageWriter}
+     * @param reader the MessageReader accepting this Message
+     * @return The return value of the MessageReader
      * @throws CloudEventRWException if something went wrong during the visit.
      * @throws IllegalStateException if the message has an unknown encoding.
      */
-    default <BV extends CloudEventWriter<R>, R> R read(MessageWriter<BV, R> visitor) throws CloudEventRWException, IllegalStateException {
+    default <BW extends CloudEventWriter<R>, R> R read(MessageWriter<BW, R> reader) throws CloudEventRWException, IllegalStateException {
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/message/MessageReader.java:83]: Again var of writer type is name reader.

File: core/src/main/java/io/cloudevents/core/message/StructuredMessageReader.java [hunk 6]
```diff
@@ -34,11 +34,15 @@
...
+     * @param <R>    the return type of the {@link StructuredMessageWriter}
+     * @param writer the writer to use to write out the message
+     * @return the return value returned by {@link StructuredMessageWriter#setEvent(EventFormat, byte[])}
      * @throws CloudEventRWException If something went wrong when
      * @throws IllegalStateException If the message is not a valid structured message
      */
-    <T> T read(StructuredMessageWriter<T> visitor) throws CloudEventRWException, IllegalStateException;
+    <R> R read(StructuredMessageWriter<R> writer) throws CloudEventRWException, IllegalStateException;
 
     default CloudEvent toEvent() throws CloudEventRWException, IllegalStateException {
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/message/StructuredMessageReader.java:47]: I see these two methods are documented in extending interface. Why?

Comment (matejvasek) [reply core/src/main/java/io/cloudevents/core/message/StructuredMessageReader.java:47]: I mean the `toEvent()` methods.

Comment (slinkydeveloper) [reply core/src/main/java/io/cloudevents/core/message/StructuredMessageReader.java:47]: I expect the user to look in this interface for this method, not in `StructuredMessageReader`. Does that makes sense?

File: core/src/main/java/io/cloudevents/core/message/MessageReader.java [hunk 7]
```diff
@@ -19,41 +19,50 @@
...
-    <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory, CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException, IllegalStateException;
+    <W extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<W, R> writerFactory, CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException, IllegalStateException;
 
     /**
-     * Visit the message as structured encoded event using the provided visitor
+     * Read the message as structured encoded message using the provided writer
      *
-     * @param visitor Structured Message visitor
+     * @param <R>    the return type of the {@link StructuredMessageWriter}
+     * @param writer Structured Message reader
```
Comment (matejvasek) [inline core/src/main/java/io/cloudevents/core/message/MessageReader.java:61]: "writer Structured Message reader"?

Comment (slinkydeveloper) [reply core/src/main/java/io/cloudevents/core/message/MessageReader.java:61]: fixed

File: api/src/main/java/io/cloudevents/rw/CloudEventWriter.java [hunk 8]
```diff
@@ -28,15 +28,16 @@
 public interface CloudEventWriter<R> extends CloudEventContextWriter {
 
     /**
-     * End the visit with a data field
+     * End the read with a data payload.
```
Comment (matejvasek) [inline api/src/main/java/io/cloudevents/rw/CloudEventWriter.java:31]: End read or write? Same below.

**Review by matejvasek — APPROVED** [review]

**Pull request merged**